yt_dlp/extractor/niconico.py

```diff
@@ -383,6 +402,41 @@ def extract_video_quality(video_quality):
             }
         }
 
+    def _extract_dms_format_for_quality(self, video_id, audio_quality, video_quality):
+
+        if not audio_quality.get('isAvailable') or not video_quality.get('isAvailable'):
+            return None
+
+        def extract_video_quality(video_quality):
+            return parse_filesize('%sB' % self._search_regex(
+                r'\| ([0-9]*\.?[0-9]*[MK])', video_quality, 'vbr', default=''))
+
+        format_id = '-'.join(
+            ['dms', remove_start(video_quality.get('id'), 'video-'), remove_start(audio_quality.get('id'), 'audio-'), 'hls'])
+
+        vid_qual_label = video_quality.get('label')
+        vid_quality = video_quality.get('bitRate')
+
+        return {
+            'url': 'niconico_dms:%s/%s/%s' % (video_id, video_quality.get('id'), audio_quality.get('id')),
+            'format_id': format_id,
+            'format_note': join_nonempty('DMS', vid_qual_label, 'hls', delim=' '),
+            'ext': 'mp4',  # Session API are used in HTML5, which always serves mp4
+            'acodec': 'aac',
+            'vcodec': 'h264',
+            'abr': float_or_none(audio_quality.get('bitRate'), 1000),
+            'vbr': float_or_none(vid_quality if vid_quality > 0 else extract_video_quality(vid_qual_label), 1000),
+            'height': video_quality.get('height'),
+            'width': video_quality.get('width'),
+            'quality': video_quality.get('qualityLevel'),
+            'protocol': 'niconico_dms',
+            'expected_protocol': "hls",
+            'http_headers': {
+                'Origin': 'https://www.nicovideo.jp',
+                'Referer': 'https://www.nicovideo.jp/watch/' + video_id,
+            }
+        }
+
```

Comment:
We can fill the "asr" field.

```suggestion
        return {
            'url': 'niconico_dms:%s/%s/%s' % (video_id, video_quality.get('id'), audio_quality.get('id')),
            'format_id': format_id,
            'format_note': join_nonempty('DMS', vid_qual_label, 'hls', delim=' '),
            'ext': 'mp4',  # Session API are used in HTML5, which always serves mp4
            'acodec': 'aac',
            'vcodec': 'h264',
            'abr': float_or_none(audio_quality.get('bitRate'), 1000),
            'vbr': float_or_none(vid_quality if vid_quality > 0 else extract_video_quality(vid_qual_label), 1000),
            'asr': float_or_none(audio_quality.get('samplingRate')),
            'height': video_quality.get('height'),
            'width': video_quality.get('width'),
            'quality': video_quality.get('qualityLevel'),
            'protocol': 'niconico_dms',
            'expected_protocol': "hls",
            'http_headers': {
                'Origin': 'https://www.nicovideo.jp',
                'Referer': 'https://www.nicovideo.jp/watch/' + video_id,
            }
        }

```

(oops, the diff is not highlighted)

Reply:
thanks!